### ISSUE TITLE
Memory Usage continues to grow

### ISSUE THREAD

**Issue by Jetcodery** [OpenLMLab/LOMO#45]
您好，我将LOMO的核心训练代码加入到了我的脚本中，可以正常训练，但是遇到了显存随训练step不断增加的情况，我在lom_trainer中打了一些查看显存占用的日志，挑选其中一个step的占用情况写明如下（我用print_memory_usage函数打印显存占用情况，结果在注释中），希望您给一些排查的方向和建议，可能是什么原因导致了这种情况的发生？：

`with tqdm.tqdm(self.train_dataloader, disable=not self.allow_print) as tqb:
    for step, batch in enumerate(tqb, start=1): # transformer.trainer:1899
        rank0_print('-' * 100)
        rank0_print(f"step {step} : {batch['input_ids'].shape}")
        self.model.train()
        print_memory_usage()  # 5844.15 MiB
        rank0_print(f"compute outs ...")
        outs = self.model(
            input_ids=batch['input_ids'].cuda(),
            attention_mask=batch['attention_mask'].cuda(),
        )
        print_memory_usage()  # 6310.91 MB
        rank0_print(f"Line 100 get_loss ...")
        loss = get_loss(outs.logits, batch['labels'], self.training_args.clip_loss_value)
        print_memory_usage()  # 6342.11 MB
        # update the learning rate
        self.global_step = self.num_steps_per_epoch * epoch + step
        self.lr = self.lr_scheduler.step(self.global_step)
        if self.training_args.clip_grad_norm is not None and self.training_args.clip_grad_norm > 0:
            rank0_print('clip_grad_norm.....')
            self.optimizer.grad_norm(loss)
            print_memory_usage()  # 6142.11 MB

            if self.optimizer.loss_scaler and self.optimizer.loss_scaler.has_overflow_serial:
                rank0_print(f"Gradient overflow, skipping step {self.global_step}")
                self.model.optimizer.get_param_coordinator(training=True).reset_step()
                print_memory_usage()
                tqb.set_postfix({'loss': loss.item()})
                if self.allow_print:
                    rank0_print({
                            'train/loss': loss.item(),
                            'train/learning_rate': self.lr,
                            'train/global_step': self.global_step,
                        })
                continue

            else:
                rank0_print('Line 129 reset_step.....')
                self.model.optimizer.get_param_coordinator(training=True).reset_step()
                print_memory_usage()  # 6142.11 MB

            # 第二次forward
            rank0_print('第二次forward.....')
            outs = self.model(
                input_ids=batch['input_ids'].cuda(),
                attention_mask=batch['attention_mask'].cuda(),
            )
            loss = get_loss(outs.logits, batch['labels'], self.training_args.clip_loss_value)
            print_memory_usage()  # 6577.68 MB

        rank0_print('fused_backward.....')
        self.optimizer.fused_backward(loss, self.lr)
        print_memory_usage()      # 6115.92 MB
        rank0_print('loss detach.....')
        loss = loss.detach()      # 6115.92 MB
        print_memory_usage()      # 6115.92 MB
        rank0_print('optimizer reset step.....')
        self.model.optimizer.get_param_coordinator(training=True).reset_step()
        print_memory_usage()      # 6115.92 MB

        tqb.set_postfix({'loss': loss.item()})
        if self.allow_print:
            rank0_print({
                    'train/loss': loss.item(),
                    'train/learning_rate': self.lr,
                    'train/global_step': self.global_step,
                })

        if self.training_args.save_strategy == 'steps' and self.global_step % self.training_args.save_steps == 0:
            self.save_model(self.global_step)

        # todo: check why memory increase
        rank0_print('del vars, empty cache.....')
        del loss, outs, batch
        torch.cuda.empty_cache()
        print_memory_usage()       # 6084.66 MB`

**Comment by Jetcodery** [comment]
@KaiLv69 @QipengGuo Looking forward to your help~

**Comment by KaiLv69** [comment]
你好，
在训练的一步中，forward之后显存占用升高是正常的，因为要存储activation。如果显存占用随着step不断增加，具体情况如果是只有前几步（或者十几步）增加，之后保持稳定也是正常的。

**Comment by Jetcodery** [comment]
您好，应该是不正常的行为，因为训练过程中，显存占用每个step会增加200M左右，从4000M的占用一直增长至24000M从而导致 OOM

**Comment by KaiLv69** [comment]
如果每个step都增加200MB，最大的可能是训练循环中有中间结果或者缓存没有被清理。也可以试试每个step手动将p.grad全部置为None来清除一下模型梯度。

**Comment by Jetcodery** [comment]
我尝试了手动清除模型梯度，但是这种情况没有发生改变，请问一下还有一些什么别的可能性么？因为关于train的核心代码我并没有做任何改变，我只改变了数据的引入方式和datacollator
